Mathlib/CategoryTheory/FiberedCategory/HomLift.lean

```diff
@@ -0,0 +1,216 @@
+  v               v
+  R -----f------> S
+```
+where the vertical arrows are given by `eqToHom` corresponding to the equalities `p(a) = R` and
+`p(b) = S`.
+
+-/
+
+universe u₁ v₁ u₂ v₂
+
+open CategoryTheory Functor Category
+
+variable {𝒮 : Type u₁} {𝒳 : Type u₂} [Category 𝒳] [Category 𝒮]
+
+/-- The proposition that an arrow a --φ--> b lifts an arrow R --f--> S in 𝒮 via p. This is
+often drawn as:
+```
+  a --φ--> b
+  -        -
+  |        |
+  v        v
+  R --f--> S
+``` -/
+class Functor.IsHomLift (p : 𝒳 ⥤ 𝒮) {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) : Prop where
+  domain_eq : p.obj a = R := by aesop_cat
+  codomain_eq : p.obj b = S := by aesop_cat
+  commsq : CommSq (p.map φ) (eqToHom domain_eq) (eqToHom codomain_eq) f
+
+namespace Functor.IsHomLift
+
+protected lemma hom_eq {p : 𝒳 ⥤ 𝒮} {R S : 𝒮} {a b : 𝒳} {f : R ⟶ S} {φ : a ⟶ b}
+    [hφ : IsHomLift p f φ] : f = eqToHom hφ.domain_eq.symm ≫ p.map φ ≫
+      eqToHom hφ.codomain_eq :=
+  ((eqToHom_comp_iff hφ.domain_eq _ _).1 hφ.commsq.w.symm)
+
+protected lemma hom_eq' {p : 𝒳 ⥤ 𝒮} {R S : 𝒮} {a b : 𝒳} {f : R ⟶ S} {φ : a ⟶ b}
+    [hφ : IsHomLift p f φ] : p.map φ = eqToHom hφ.domain_eq ≫ f ≫
+      eqToHom hφ.codomain_eq.symm := by
+  rw [← assoc, ← comp_eqToHom_iff hφ.codomain_eq _ _]
+  exact hφ.commsq.w
+
+lemma eq_of_isHomLift {p : 𝒳 ⥤ 𝒮} (a b : 𝒳) {f : p.obj a ⟶ p.obj b} {φ : a ⟶ b}
+    [hφ : IsHomLift p f φ] : f = p.map φ := by
+  simpa only [eqToHom_refl, comp_id, id_comp] using hφ.hom_eq
+
+/-- For any arrow `φ : a ⟶ b` in `𝒳`, `φ` lifts the arrow `p.map φ` in the base `𝒮`-/
+instance self (p : 𝒳 ⥤ 𝒮) {a b : 𝒳} (φ : a ⟶ b) : IsHomLift p (p.map φ) φ where
+  commsq := ⟨by simp only [eqToHom_refl, comp_id, id_comp]⟩
+
+instance (p : 𝒳 ⥤ 𝒮) (a : 𝒳) : IsHomLift p (𝟙 (p.obj a)) (𝟙 a) :=
+  p.map_id _ ▸ self p (𝟙 a)
+
+@[simp]
+protected lemma id {p : 𝒳 ⥤ 𝒮} {R : 𝒮} {a : 𝒳} (ha : p.obj a = R) : IsHomLift p (𝟙 R) (𝟙 a) :=
+  ha ▸ (p.map_id _ ▸ self p (𝟙 a))
+
+instance comp {p : 𝒳 ⥤ 𝒮} {R S T : 𝒮} {a b c : 𝒳} {f : R ⟶ S}
+    {g : S ⟶ T} {φ : a ⟶ b} {ψ : b ⟶ c} [hφ : IsHomLift p f φ]
+    [hψ : IsHomLift p g ψ] : IsHomLift p (f ≫ g) (φ ≫ ψ) where
+  domain_eq := hφ.1
+  codomain_eq := hψ.2
+  commsq := (p.map_comp _ _).symm ▸ CommSq.horiz_comp hφ.3 hψ.3
+
+/-- If `φ : a ⟶ b` and `ψ : b ⟶ c` lift `𝟙 S`, then so does `φ ≫ ψ` -/
+instance lift_id_comp {p : 𝒳 ⥤ 𝒮} {R : 𝒮} {a b c : 𝒳} {φ : a ⟶ b} {ψ : b ⟶ c}
+    [hφ : IsHomLift p (𝟙 R) φ] [IsHomLift p (𝟙 R) ψ] : IsHomLift p (𝟙 R) (φ ≫ ψ) :=
+  comp_id (𝟙 R) ▸ hφ.comp
+
+/-- If `φ : a ⟶ b` lifts `f` and `ψ : b ⟶ c` lifts `𝟙 T`, then `φ  ≫ ψ` lifts `f` -/
+lemma comp_lift_id {p : 𝒳 ⥤ 𝒮} {R S T : 𝒮} {a b c : 𝒳} {f : R ⟶ S}
+    {φ : a ⟶ b} [hφ : IsHomLift p f φ] {ψ : b ⟶ c} [hψ : IsHomLift p (𝟙 T) ψ] :
+    IsHomLift p f (φ ≫ ψ) where
+  domain_eq := hφ.domain_eq
+  codomain_eq := by rw [hψ.codomain_eq, ← hψ.domain_eq, hφ.codomain_eq]
+  commsq := ⟨by simp only [map_comp, hψ.hom_eq', id_comp, eqToHom_trans, assoc, hφ.3.1]⟩
```

Comment:
The idea of using a typeclass is that the parameter should no longer need a name `hφ`. You should state lemmas like `domain_eq_of_isHomLift` which would take `p`, `f`, `φ`, and `[IsHomLift p f φ]` as parameters and would produce the expected equality. Then, instead of `hφ.domain_eq`, you would write `domain_eq_of_isHomLift p f φ`.

Reply:
Should I never do `hφ.domain_eq`? I get that it adds automation, so I should definitely add the simp lemma `domain_eq_of_isHomLift`, but I fell that  `hφ.domain_eq` is a lot shorter than `domain_eq_of_isHomLift p f φ`

Reply:
You may try to find a shorter name instead of `domain_eq_of_isHomLift`: it could be a very short name inside a suitable namespace!